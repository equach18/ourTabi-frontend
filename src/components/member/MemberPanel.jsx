import { useState, useContext } from "react";
import UserContext from "../../context/UserContext";
import MemberCard from "./MemberCard";

function MemberPanel({ isTripCreator, members }) {
  const { friends } = useContext(UserContext);
  const [isAdding, setIsAdding] = useState(false);
  const eligibleFriends = friends.filter(
    (friend) => !members.some((member) => member.userId === friend.userId)
  );

  return (
    <div className="h-full flex flex-col bg-white rounded-xl shadow overflow-hidden">
      <div className="flex justify-between items-center p-4">
        <h2 className="text-xl font-semibold">Members</h2>
        {isTripCreator && (
          <button
            onClick={() => setIsAdding(!isAdding)}
            className="text-sm bg-emerald-600 text-white px-3 py-1 rounded hover:bg-emerald-700"
            aria-label={isAdding ? "Back to members list" : "Show eligible friends"}
          >
            {isAdding ? (
              <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={1.5}
                stroke="currentColor"
                className="size-5"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  d="M9 15 3 9m0 0 6-6M3 9h12a6 6 0 0 1 0 12h-3"
                />
              </svg>
            ) : (
              <svg
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth={1.5}
                stroke="currentColor"
                className="size-5"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  d="M18 7.5v3m0 0v3m0-3h3m-3 0h-3m-2.25-4.125a3.375 3.375 0 1 1-6.75 0 3.375 3.375 0 0 1 6.75 0ZM3 19.235v-.11a6.375 6.375 0 0 1 12.75 0v.109A12.318 12.318 0 0 1 9.374 21c-2.331 0-4.512-.645-6.374-1.766Z"
                />
              </svg>
            )}
          </button>
        )}
      </div>

      <div className="flex-1 overflow-y-auto px-4 py-2 space-y-3">
        {!isAdding ? (
          members.length > 0 ? (
            members.map((member) => (
              <MemberCard
                key={member.userId}
                member={member}
                isTripCreator={isTripCreator}
              />
            ))
          ) : (
            <p className="text-gray-500">No members yet.</p>
          )
        ) : eligibleFriends.length > 0 ? (
          eligibleFriends.map((friend) => (
            <MemberCard key={friend.userId} member={friend} isEditing />
          ))
        ) : (
          <p className="text-gray-500">No eligible friends to add.</p>
        )}
      </div>
    </div>
  );
}

export default MemberPanel;
