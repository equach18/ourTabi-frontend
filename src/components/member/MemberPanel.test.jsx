import { screen, fireEvent } from "@testing-library/react";
import MemberPanel from "./MemberPanel";
import {
  renderWithAllProviders,
  testUser1,
  testUser2,
  testUser3,
  defaultMembers,
} from "../../test-helpers/_testCommons";

const testUserContext = {
  currentUser: testUser1,
  friends: [
    { ...testUser2, userId: testUser2.id },
    { ...testUser3, userId: testUser3.id },
  ],
  incomingRequests: [],
  sentRequests: [],
};

describe("MemberPanel", () => {
  it("matches snapshot for default", () => {
    const { asFragment } = renderWithAllProviders(
      <MemberPanel isTripCreator={false} members={defaultMembers} />,
      { userContext: testUserContext }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders member list on default", () => {
    renderWithAllProviders(
      <MemberPanel isTripCreator={false} members={defaultMembers} />,
      { userContext: testUserContext }
    );

    expect(screen.getByText("@testUser1")).toBeInTheDocument();
    expect(screen.getByText("@testUser2")).toBeInTheDocument();
  });

  it("shows msg when there are no members", () => {
    renderWithAllProviders(<MemberPanel isTripCreator={false} members={[]} />, {
      userContext: testUserContext,
    });

    expect(screen.getByText(/no members yet/i)).toBeInTheDocument();
  });

  it("Shows eligible friends when add button clicked", () => {
    renderWithAllProviders(
      <MemberPanel isTripCreator={true} members={defaultMembers} />,
      { userContext: testUserContext }
    );

    const addBtn = screen.getByRole("button", {
      name: /show eligible friends/i,
    });
    fireEvent.click(addBtn);

    expect(screen.getByText("@testUser3")).toBeInTheDocument();
  });

  it("Shows msg when there are no eligible friends to add", () => {
    renderWithAllProviders(
      <MemberPanel isTripCreator={true} members={defaultMembers} />,
      {
        userContext: {
          currentUser: testUser1,
          friends: [{ ...testUser2, userId: testUser2.id }],
          incomingRequests: [],
          sentRequests: [],
        },
      }
    );

    const addBtn = screen.getByRole("button", {
      name: /show eligible friends/i,
    });
    fireEvent.click(addBtn);

    expect(screen.getByText(/no eligible friends to add/i)).toBeInTheDocument();
  });
});
