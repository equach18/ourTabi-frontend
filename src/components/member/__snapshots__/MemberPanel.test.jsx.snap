// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MemberPanel > matches snapshot for default 1`] = `
<DocumentFragment>
  <div
    class="h-full flex flex-col bg-white rounded-xl shadow overflow-hidden"
  >
    <div
      class="flex justify-between items-center p-4"
    >
      <h2
        class="text-xl font-semibold"
      >
        Members
      </h2>
    </div>
    <div
      class="flex-1 overflow-y-auto px-4 py-2 space-y-3"
    >
      <div
        class="flex items-center justify-between bg-gray-50 p-3 rounded-lg shadow-sm"
      >
        <div
          class="flex items-center gap-3"
        >
          <img
            alt="testUser1's avatar"
            class="w-10 h-10 rounded-full object-cover border"
            src="/src/assets/profilePic.jpg"
          />
          <div
            class="text-sm"
          >
            <p
              class="font-semibold"
            >
               
            </p>
            <p
              class="text-gray-500"
            >
              @testUser1
            </p>
          </div>
        </div>
        <div
          class="flex gap-2"
        />
      </div>
      <div
        class="flex items-center justify-between bg-gray-50 p-3 rounded-lg shadow-sm"
      >
        <div
          class="flex items-center gap-3"
        >
          <img
            alt="testUser2's avatar"
            class="w-10 h-10 rounded-full object-cover border"
            src="testUser2.png"
          />
          <div
            class="text-sm"
          >
            <p
              class="font-semibold"
            >
               
            </p>
            <p
              class="text-gray-500"
            >
              @testUser2
            </p>
          </div>
        </div>
        <div
          class="flex gap-2"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
